swiftglue/registeraccess.c

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// These are accessors for each supported CPU to get the register arguments.
+// These are used because swift puts implicit arguments in registers including
+// type metadata and protocol witness tables.
+
+// supported CPUs:
+// __x86_64
+// __arm64
+// __arm
+// i386
+// we could probably further distinguish armv7, armv7k, and armv7s, but
+// I don't imagine that we'll need the extensions since all we're doing is
+// register juggling
+
+void swiftAsmArg0 ()
+{
+#if __x86_64
+	__asm("mov %rdi, %rax");
+#elif __arm64
+	__asm("mov x0, x0");
+#elif __arm
+	__asm("mov r0, r0");
+#elif i386
+	asm {
+		mov eax, dword ptr [ebp + 12]
+	}
+#else
+#error("Unknown CPU type for register access");
+#endif
+}
```

Comment:
Wouldn't this work just as well?

```c
void *swiftArg0 (void *p0)
{
    return p0:
}

void *swiftArg1 (void *p0, void *p1)
{
    return p1;
}
...
```

And it's bitcode-compatible!

See: https://gist.github.com/rolfbjarne/6181be7133c129670196c43619ac744d

Reply:
Oh this is hilarious - lie about the signature in the export vis the implementation! Have verified that it works.